### DISCUSSION TITLE
Rethinking CKAN's Unit Test approach.

### DISCUSSION THREAD

**Discussion by pdelboca** [ckan/ckan#6373] [Ideas]
Hey all! :wave:  

I have been spending a few days working on Unit Test for a project and I would like to open the floor for a discussion about our approaches and experiences. The goal is to discuss and maybe, if needed, come up with improvements to our tools to do Unit Testing.

The main issue I found is that **our unit tests are not really units**. When trying to write big suits of Unit Tests it gets hard to keep each test/class/module independent.

### Issue 1:
Our main factories write to the database and don't clean it after execution. Even when we have `clean_db` fixture, each test is resposible for cleaning it's environment before executing in case other tests messed up with the database. In a Unit Testing approach each test should be responsible for cleaning their own mess.

### Issue 2:
Adding tables to the database have the same side effect than the previous issue. We have the fixture `with_plugins` to load them while testing but if we need to initialize the database to add tables and columns then we depend on other tests to call `reset_db` so they have a clean DB before run.

### Issue 3:
Issue 1 and 2 cause an indiscriminate use of fixtures like `clear_db` that ends up in an overhead of the database. Also, with `clean_db` we are deleting and initializing the database when most of the time we just need to cleanup the tables.

### Issue 4:
We are using fixtures only to write tools that allows us to clean/modify the environment and relying on factories for ORM objects. A better approach would be to use fixtures to create ORM objects as well. (something similar to [django-pytest](https://pytest-django.readthedocs.io) maybe?) Some work is being done in: https://github.com/ckan/ckan/pull/6335

Overall the experience to write tests it's okay but it gets messy when the implementations starts interacting with several extensions.

### Possible solutions:

The approach for unit testing and fixture building should be to isolate each function and to cleanup the environment after each execution.

### Example 1: Refactor `clear_db` fixture

```python
from ckan.model import repo

@pytest.fixture(autouse=True)
def clearn_db():
    yield
    repo.delete_all()
```
This fixture will delete all the data from the tables once the execution of the test finishes

### Example 2: Writing some good practices for unit test different implementations

- What's a good approach to test an implementation that constantly interacts with extensions that extends the database model? 
- Should we add all the extensions to the `test.ini` file an initialize the database once before executing the test or create fixtures for each extension to set/clean the database? (Maybe each extensions can provide a fixtures to encapsulate implementation details to other users?)

Let me know your thoughts!

Thanks!

**Comment by smotornyuk** [comment]
> When trying to write big suits of Unit Tests it gets hard to keep each test/class/module independent.

What exactly prevents your tests from being independent? 
```
@pytest.mark.ckan_config("ckan.plguins", "my_plugin")
@pytest.mark.usefixtures("with_plugins", "clean_db", "clean_index")
```
and your test is absolutely independent from every other test. And if it's too wordy for you, just create `independent` fixture, that applies changes above

## Issue 1: 
>  In a Unit Testing approach each test should be responsible for cleaning their own mess.

Do not agree. The test is responsible only for correct assertions. Of course, if the test has **messed up** the system, the system must be restored. But a *dirty database* it's just an artifact, that can be investigated after tests, it's not a mess.
 At the same time, **clean database** it's an initial requirement, so it has more sense to verify that all the requirements are satisfied before the test starts(i.e, use setUp stage) than to rely on other tests to clean their mess. If you rely on other tests, your test-suite turns into a fragile structure where all the consequent tests will fail if one of them accidentally won't clean the database

## issue 2:

Just as you said, it's the same :)


## Issue 3:

> with clean_db we are deleting and initializing the database when most of the time we just need to clean up the tables.


it looks like you haven't checked the code :) 

> Issue 1 and 2 cause an indiscriminate use of fixtures like clear_db 

Just use it deliberately)


## Issue 4

Even though I like fixtures and I suggested using them, I don't agree that it's **better**. It's just a matter of taste and style, but I prefer to have a choice:)


## Example 1: Refactor clear_db fixture

So, what is actual change? Instead of cleaning up **before** we'll do clean-up **after** the test. Neither from the user's point of view nor from a performance point it will improve the situation. But, if we'll do cleaning-up **before** the test(like we are doing right now), we can remove the `clean_db` fixture from the tests that don't mind against existing data. For example, tests that are creating a dataset and modifying this particular dataset - will work if there is some existing data in DB. In that way there will be fewer clean-ups and tests will become faster.

Instead, I suggest using an extra fixture, `with_db`, that makes sure that DB is initialized, but does not try to remove existing data. This means that the very first test with the `with_db` fixture will initialize the database and all the following tests will get no-op for this fixture. 


## Example 2: Writing some good practices for unit test different implementations

+1
BTW, [we have a fixture for applying DB migrations from extensions](https://github.com/ckan/ckan/pull/6139). Personally, I'm redefining `clean_db` in my extensions with something like 
```
def clean_db(reset_db, migrate_db_for):
    reset_db()
    migrate_db_for(my_extension)
```

But as I currently working on speeding up CKAN tests (actually I'm waiting till #6335 merged, and then I'll be able to continue), I can suggest a better way in future

**Comment by pdelboca** [comment]
Hey! Thanks for answering :)

## Issue 1 & 2

The design principle of Unit Testing approach is that each test should clean up the database after execution. As per pytest documentation:

> You can think of a test as being broken down into four steps:
>  1. Arrange
>  2. Act
>  3. Assert
>  4. Cleanup
>
> **Cleanup** is where the test picks up after itself, so other tests aren’t being accidentally influenced by it.

But our current approach is: `Cleanup -> Arrange -> Act -> Assert.`. Also, in our approach the cleanup is if needed, so to answer that question I need to be aware of what other tests are doing to the database.

We both agree that we do not want our test-suite turns into a fragile structure where all the consequent tests will fail if one of them accidentally won't clean the database. **It's just a matter of who is responsible for cleaning**. pytest approach (and unit testing frameworks) assign that responsibility to the test creating/editing the environment (teardown methods). Which makes sense since: why I should introduce in my test B a cleanup for logic that other methods may have introduced or not?

We do it in in fixtures like `with_plugins` and `ckan_config`. So I think we should also do it with Database objects and other fixtures [as pytest also recommends](https://docs.pytest.org/en/6.2.x/fixture.html#teardown-cleanup-aka-fixture-finalization):

> When we run our tests, we’ll want to make sure they clean up after themselves so they don’t mess with any other tests (and also so that we don’t leave behind a mountain of test data to bloat the system). Fixtures in pytest offer a very useful teardown system, which allows us to define the specific steps necessary for each fixture to clean up after itself.

And an example of their documentation:
``` python

@pytest.fixture
def sending_user(mail_admin):
    user = mail_admin.create_user()
    yield user
    admin_client.delete_user(user)
```

## Issue 3

The code for `clean_db` became difficult to follow as it: `clean_db -> reset_db ->  reset_db -> rebuild_db -> init_db`. But `init_db` inits only if. 

Re-checking my examples I noticed that it calls `delete_all()` if tables exists but the first execution it calls `clean_db` (the flag `tables_created_and_initialised` is only set on `call_db`) and that created my confusion.

Note: is it safe to do a refactor and only use `are_tables_created` method here?

The problem with using deliberately `clean_db` is that to answer that question I need to know if previous tests made something with my database. So what's the criteria for using it? We could say that the criteria will be: **if you need to use the database, clean it** but then:
 - I can be cleaning the database with no reason (because no previous test used it)
 - I can add `clean_db` to test B because previous test A left the database dirty. But if in a follow up refactor I remove test A, I need to also remember to remove `clean_db` from test B.
 - If test B is not using `clean_db` can fail if I introduce a new test C before it or if I refactor Test A then it may break test B.

Same applies if suddenly I add a new extension to my work and some tests now requires database migrations to work. Would I add `clean_db` to all the other tests to restore the database schema? (Or just to the first one next in the chain because that will work for the following ones)

Again, it is a matter on who clean but it is not a good software pattern (and it is not expected) to make classes/code deal with other's mess. (Unless you're writting a garbage collector of course ;))

### New fixture: with_db

IIUC I like the overall idea, it's something similar to what `pytest-django` does. A conservative approach to database connections can be good since it will make the user more aware of it's use.

## Example 2

Thanks for the tip on the DB migration, I will do some refactors in my code :)

**Reply by smotornyuk** [reply]
> I need to be aware of what other tests are doing to the database.

No, you don't. If your test expects a clean database - you ask for it. If it doesn't - you won't ask for it. And no matter, whether there are other tests, nothing will change. Because `clean_db` it's not an verb(cleanUp stage), it's a noun(arrangement stage)

> It's just a matter of who is responsible for cleaning

All the time the one who created the mess is responsible for cleaning. But data in DB - it's not a mess, it's an artifact. If you run the single test - you can investigate it. If you run multiple db-related tests - every test starts in a clean new world without any artifacts, as long as `clean_db` requirement for the test is satisfied.

> The problem with using deliberately clean_db is that to answer that question I need to know if previous tests made something with my database.
 
That's where you are wrong. You don't have to know anything about other tests. If your test can fail on non-empty DB it is either incorrect or must be marked as a test that requires a clean DB. You should do no assumptions about the environment because that makes your tests fragile. 

`clean_db` - it's not ad **CleanUp**. it's `Arrange`. I.e, we have this workflow:

* Arrange: clean_db arranges database for the user. If the database does not exist - it's created. If it's empty nothing happens. If it's dirty - the fixture arranges everything so that this dirty state won't affect your tests. 
* Act 
*  Assert
* CleanUp - in our case, these `with_plugins`/`ckan_config` do the clean-up. We can add an extra fixture that will remove the data from DB, it's not a problem, if things that are left in DB must be removed. For example, if you've created extra tables, or created unremovable (magic of foreign keys) objects. You can actually remove all the data created by your tests, I don't mind. But `clean_db` does not remove data for you, it **prepares** database for you. If you want a fixture that will remove the data - create it.

You are interpreting these recommendations literally. Yes, it has some sense to move `clean_db` into the CleanUp phase, but it means every test will do cleaning, even if it's not required(and it will make reducing CKAN core's testing impossible(20 minutes at the moment) - that's the main reason why I'm confronting you right now :) ). You see, cleaning the database is required not because of some obscure reasons. The only purpose of it - not to break other tests. As long as leaving DB in a dirty state won't affect any other test - it has no value. 


> t is not a good software pattern (and it is not expected) to make classes/code deal with other's mess.

So, once again - our tests do not work with the mess from other tests. Our tests require a clean database. You know about underlying logic and that's why you are interpreting this step as cleaning up. But it's preparation and you should forget about other tests.

> I can be cleaning the database with no reason (because no previous test used it)

and it's a cheap operation because you removing data from empty tables(actually, it can be cheaper, but it requires a bit of work).

> I can add clean_db to test B because previous test A left the database dirty. But if in a follow-up refactor I remove test A, I need to also remember to remove clean_db from test B.

there should be no such thing as `previous test` in your way of thinking. You are making tests dependent on each other. If your test B requires a clean database - add this fixture, no matter are there any other tests. It's the responsibility of the fixture itself to be fast and smart and avoid unnecessary work. As long as the order of test execution can affect the result, you are doing things in the wrong way. You should add `clean_db` not because B runs after A, but only if `B needs clean database`. There is no A exists when you thinks about B test

> If test B is not using clean_db can fail if I introduce a new test C before it or if I refactor Test A then it may break test B.

If your test can fail after adding another test - your tests are bad. Once again, let's assume you have three tests - A, B, C. All of the **do cleanUp**(remove data when they finished). If they run randomly and test A for some reason failed to clean the database. Problem not in the fixture, because it that case all tests weren't able to clean up the database. Something wrong with this particular test. 
When it's executed first - all three tests fail. When it's executed last - it's the only failed test. This makes your tests non-deterministic.

Now let's consider using `clean_db` as an arrangement. If one test fails(once again, the problem with the test, not with the fixture), all the other tests will pass, no matter what was the order.

All the recommendations must never be blindly followed. Because sometimes, a person can misinterpret something. in our case you looking assumes that `clean_db` is a cleanUp, but for us, it's an Arrangement. 

PS once again, I don't mind your interpretation. Actually, I think in the same way. But we need a current workflow in order to make things faster. That's why we have to look at it from the different angle

**Comment by smotornyuk** [comment]
I think we need more fixtures with different responsibilities. Like (of course, names can be better :smile:  ):

* with_db - just initialized database. If it doesn't exist, let's create it. If it exists, do nothing
* clean_db - a database that is clean
* do_not_leave_a_mess - removes all the data from DB after the test
* etc.

Right now we have a lack of expressiveness and that makes tests more stiff and uncertain.

**Comment by pdelboca** [comment]
Clarification: Let's not take my `clear_db` example as literal, that's not the implementation I want either. It was an example to explain my key point that every fixture/test that writes to the database, should be responsible for cleaning the data. And agree that we need more expressiveness in our fixtures.

Yes, databases are **artifacts** and I agree with some kind of feature to explicitly request for it (like `with_db`). In fact, I would fail if the user is trying to access the DB without requesting for it. What I disagree with are fixtures and tests that do changes to the database and don't take responsibility to cleanup after using it. (and it's no that I disagree, it's the common practice of Unit Testing to clean data before execution) Databases are artifacts, rows in the database are not.

As pytest documentation suggest, if a fixture creates a User (or a connection), the fixture should remove it after using it. What's the goal of creating `user_a` to test in Test A and leave it there? What's the goal of adding a new table/relation and leave it there?

Some example of what it's confusing to me:

```python
class TestTests:
    def test_a(self):
        factories.Sysadmin()

class TestTests2:
    def test_b(self):
        factories.Sysadmin()
```

The previous example will:
 - Run okay on the first execution: `pytest --ckan-ini=test.ini tests/test_tests.py`
 - Fail on the second execution due to `duplicate key value violates unique constraint "user_name_key"` 

To fix it I need to execute `ckan -c ./test.ini db clean && ckan -c ./test.ini db init` before executing the tests or add `clean_db` fixture to it. So I need to:

```python
@pytest.mark.usefixtures('clean_db')
class TestTests:
    def test_a(self):
        factories.Sysadmin()

@pytest.mark.usefixtures('clean_db')
class TestTests2:
    def test_b(self):
        factories.Sysadmin()
```

That would run okay all the times but it requires running a `delete_all()` twice. So we are cleaning all tables in the database for one row in sysadmin. Also you need to clean the database either in the first test (because the last test running in the suit leaves it dirty) or execute `db init` before executing `pytest`.

If we do something like:
```python
@pytest.mark.usefixtures('with_db')
@pytest.mark.usefixtures('do_not_leave_a_mess')
class TestTests:
    def test_a(user_factory):
        user = user_factory(name="xxx")

@pytest.mark.usefixtures('with_db')
@pytest.mark.usefixtures('do_not_leave_a_mess')
class TestTests2:
    def (user_factory):
        user = user_factory(name="xxx")
```

We have a more expressive fixture but again, `do_not_leave_a_mess` means `clean everything because i don't know what's the status of the database after the test`. Maybe new factories are more intelligent and don't throw unique constraints but we are still having messy data in the database + a garbage-collector-like fixture.

A more clean approach for me will be something like:
```python
@pytest.mark.usefixtures('with_db')
class TestTests:
    def test_a(user_factory):
        user = user_factory(name="xxx")

@pytest.mark.usefixtures('with_db')
class TestTests2:
    def (user_factory):
        user = user_factory(name="xxx")

@pytest.fixture
def user_factory():
    _user = User(name="Susan")
    yield _user
    _user.delete() # or something to clean it from the database
```

This way:
 - We have clean approach for fixtures that's what most users that comes from other project's will expect
 - We are more aligned to pytest framework
 - We also have cleaner deletes on the database. We have concrete deletes instead of iteratively delete all from everything for all the tests accessing the database. (Should be better for performance)

**Reply by smotornyuk** [reply]
Now i agree, but..  Not deleting is even faster,  than deleting:)

As first iteration your approach is good.  But it leaves us with overcomplicated fixtures and without an ability to check database records after the fail. And it won't give big speed boost really,  i have tested it when started pytest migration :(

And instead, if we use random data and as long as your sysadmin-test(and any other test as well)  won't fail even if you won't clean the DB, does this mandatory cleanup worth two drawbacks mentioned above? (just let's assume that we can guarantee unique data and ignore the case when you want to use some particular non-unique id)

**Reply by smotornyuk** [reply]
PS i'm focused only on CKAN core tests in this argument, as fixtures for external usage does not worth discussion - it's always a good idea . If you have time to create PR with suggested fixtures for extensions,  i definitely would like them to be merged right away
